File: src/core/cloudflare-actions.ts
import {
  cloudflareRequest,
  type CloudflareEnvVar,
  type CloudflareRequestAuth,
} from "./cloudflare-manager.js";

/**
 * Write-capable Cloudflare Pages operations, deliberately separate from the
 * read-safe {@link ./cloudflare-manager.js CloudflareManager} — the same split
 * as `git-manager` / `git-actions` and `db-peek` / `db-write`.
 *
 * Everything here changes what the internet is serving, so none of it is
 * exposed as an MCP tool: these are reached only from the dashboard's own
 * routes, where a human clicked the button.
 *
 * Still intentionally excludes the irreversible ones — Cloudflare's
 * `DELETE .../deployments/{id}` and project deletion — which would need their
 * own guarded surface.
 */

/** The only two environments Pages has. A third would be silently created by a PATCH. */
export const CLOUDFLARE_ENVIRONMENTS = ["production", "preview"] as const;
export type CloudflareEnvironment = (typeof CLOUDFLARE_ENVIRONMENTS)[number];

export class CloudflareActions {
  constructor(
    private readonly auth: CloudflareRequestAuth,
    private readonly accountId: string,
  ) {}

  /**
   * Rebuild an existing deployment.
   *
   * Cloudflare calls this "retry" and mints a new deployment id, inheriting the
   * original's branch, commit and environment — so, unlike Vercel's redeploy,
   * nothing has to be carried through by the caller to keep a production retry
   * in production.
   */
  async retry(
    projectName: string,
    deploymentId: string,
  ): Promise<{ id: string; url: string | null }> {
    const created = await this.request<{ id?: string; url?: string | null }>(
      `${this.projectPath(projectName)}/deployments/${encodeURIComponent(deploymentId)}/retry`,
      { method: "POST" },
    );
    return { id: created.id ?? "", url: created.url ?? null };
  }

  /**
   * Point the project's production alias at an earlier deployment without
   * rebuilding it. Cloudflare records this as a rollback, and it takes effect
   * immediately.
   */
  async rollback(
    projectName: string,
    deploymentId: string,
  ): Promise<{ id: string; url: string | null }> {
    const rolled = await this.request<{ id?: string; url?: string | null }>(
      `${this.projectPath(projectName)}/deployments/${encodeURIComponent(deploymentId)}/rollback`,
      { method: "POST" },
    );
    return { id: rolled.id ?? deploymentId, url: rolled.url ?? null };
  }

  /**
   * Add a variable. `secret` (Vercel's "encrypted") is the default, and its
   * value never reads back over the API afterwards.
   */
  createEnv(
    projectName: string,
    input: { key: string; value: string; environments: string[]; secret?: boolean },
  ): Promise<CloudflareEnvVar> {
    const type = input.secret === false ? "plain_text" : "secret_text";
    const patch: EnvPatch = {};
    for (const environment of requireEnvironments(input.environments)) {
      patch[environment] = { [input.key]: { type, value: input.value } };
    }
    return this.patchEnv(projectName, input.key, patch);
  }

  /**
   * Change a variable's value and/or the environments it applies to.
   *
   * Adding an environment needs a value: Cloudflare stores the variable
   * separately per environment and never hands back a secret's value, so there
   * is nothing to copy across. Saying that is better than writing an empty
   * string into the new environment.
   */
  async updateEnv(
    projectName: string,
    key: string,
    input: { value?: string; environments?: string[] },
  ): Promise<CloudflareEnvVar> {
    const current = await this.currentEnv(projectName, key);
    const wanted = input.environments
      ? requireEnvironments(input.environments)
      : (current.environments as CloudflareEnvironment[]);

    const patch: EnvPatch = {};
    for (const environment of wanted) {
      if (input.value !== undefined) {
        patch[environment] = { [key]: { type: current.type, value: input.value } };
      } else if (!current.environments.includes(environment)) {
        throw new Error(
          `Adding "${key}" to ${environment} needs a value — Cloudflare stores it separately per environment.`,
        );
      }
    }
    for (const environment of current.environments) {
      if (!wanted.includes(environment as CloudflareEnvironment)) {
        patch[environment] = { [key]: null };
      }
    }
    return this.patchEnv(projectName, key, patch);
  }

  /** Remove a variable from every environment it is set in. */
  async deleteEnv(projectName: string, key: string): Promise<void> {
    const current = await this.currentEnv(projectName, key);
    const patch: EnvPatch = {};
    for (const environment of current.environments) patch[environment] = { [key]: null };
    await this.patchProject(projectName, patch);
  }

  /**
   * The one write to a project's variables.
   *
   * Cloudflare merges the `env_vars` map it is sent — a key set to `null` is
   * deleted, and keys the patch does not mention are untouched — so this never
   * has to read-modify-write the whole map and cannot clobber a variable added
   * between the read and the write.
   */
  private async patchEnv(
    projectName: string,
    key: string,
    patch: EnvPatch,
  ): Promise<CloudflareEnvVar> {
    const updated = await this.patchProject(projectName, patch);
    const found = updated.find((entry) => entry.key === key);
    if (!found) throw new Error(`Cloudflare did not report "${key}" after the update.`);
    return found;
  }

  private async patchProject(projectName: string, patch: EnvPatch): Promise<CloudflareEnvVar[]> {
    const deploymentConfigs: Record<string, { env_vars: EnvPatchEntry }> = {};
    for (const [environment, envVars] of Object.entries(patch)) {
      if (envVars) deploymentConfigs[environment] = { env_vars: envVars };
    }
    const project = await this.request<RawPatchedProject>(this.projectPath(projectName), {
      method: "PATCH",
      body: { deployment_configs: deploymentConfigs },
    });
    return mergeEnvVars(project.deployment_configs);
  }

  private async currentEnv(projectName: string, key: string): Promise<CloudflareEnvVar> {
    const project = await this.request<RawPatchedProject>(this.projectPath(projectName));
    const found = mergeEnvVars(project.deployment_configs).find((entry) => entry.key === key);
    if (!found) throw new Error(`No variable named "${key}" on this project.`);
    return found;
  }

  private projectPath(projectName: string): string {
    return `/accounts/${this.accountId}/pages/projects/${encodeURIComponent(projectName)}`;
  }

  private request<T>(path: string, opts?: { method?: string; body?: unknown }): Promise<T> {
    return cloudflareRequest<T>(this.auth, path, opts);
  }
}

type EnvPatchEntry = Record<string, { type: string; value: string } | null>;
type EnvPatch = Record<string, EnvPatchEntry | undefined>;

interface RawPatchedProject {
  deployment_configs?: Record<
    string,
    { env_vars?: Record<string, { type?: string; value?: string } | null> | null } | null
  > | null;
}

/**
 * Rejects anything that is not one of Pages' two environments. Vercel's env
 * dialog offers `development` as well, and a PATCH naming it would quietly
 * create a third deployment config that nothing ever reads.
 */
function requireEnvironments(environments: string[]): CloudflareEnvironment[] {
  const invalid = environments.filter(
    (entry) => !CLOUDFLARE_ENVIRONMENTS.includes(entry as CloudflareEnvironment),
  );
  if (invalid.length > 0) {
    throw new Error(
      `Cloudflare Pages only has ${CLOUDFLARE_ENVIRONMENTS.join(" and ")} environments, not ${invalid.join(", ")}.`,
    );
  }
  return environments as CloudflareEnvironment[];
}

/** The read side's flattening, applied to whatever a write handed back. */
function mergeEnvVars(
  configs: RawPatchedProject["deployment_configs"],
): CloudflareEnvVar[] {
  const merged = new Map<string, CloudflareEnvVar>();
  for (const [environment, config] of Object.entries(configs ?? {})) {
    for (const [key, entry] of Object.entries(config?.env_vars ?? {})) {
      if (!entry) continue;
      const type = entry.type === "plain_text" ? "plain_text" : "secret_text";
      const existing = merged.get(key);
      if (existing) {
        existing.environments.push(environment);
        if (type === "secret_text") existing.type = "secret_text";
        continue;
      }
      merged.set(key, { key, environments: [environment], type, value: entry.value });
    }
  }
  return [...merged.values()].sort((a, b) => a.key.localeCompare(b.key));
}

/**
 * Cloudflare's action vocabulary, mapped onto the names the shared UI already
 * knows. "retry" is Cloudflare's word for what Vercel calls a redeploy, and
 * there is no cancel endpoint for a Pages build at all — so the manifest is
 * two actions, not four, and the deployment row hides what it cannot do.
 */
export const CLOUDFLARE_ACTIONS = ["redeploy", "rollback"] as const;
export type CloudflareActionName = (typeof CLOUDFLARE_ACTIONS)[number];

/** Actions that change what production serves — the UI confirms these first. */
export const PRODUCTION_AFFECTING_ACTIONS: ReadonlySet<CloudflareActionName> = new Set([
  "rollback",
]);

File: crates/nomoreide-core/src/cloudflare_actions.rs
//! Write-capable Cloudflare Pages operations, deliberately separate from the
//! read-safe `cloudflare_manager` — the same split as `git_manager` / git write
//! commands, and the Rust counterpart of `src/core/cloudflare-actions.ts`.
//!
//! Everything here changes what the internet is serving, so none of it is
//! exposed to an agent: these are reached only from the dashboard's own routes,
//! where a human clicked the button.
//!
//! Still intentionally excludes the irreversible ones — Cloudflare's
//! `DELETE .../deployments/{id}` and project deletion — which would need their
//! own guarded surface.
//!
//! **A variable is not a record here, it is a patch.** Pages has no env-var
//! endpoint at all: variables live inside the project's `deployment_configs`,
//! one map per environment, so every write below is a `PATCH` of the project.
//! Cloudflare *merges* the map it is sent — a key set to `null` is deleted and
//! keys the patch does not mention are untouched — which is what lets these
//! avoid a read-modify-write that could clobber a variable added in between.

use serde_json::{Map, Value};

use crate::cloudflare_manager::{
    merge_env_vars, request_json, CloudflareApiError, CloudflareEnvVar, PLAIN_TEXT, SECRET_TEXT,
};
use crate::providers::deploy::CreatedDeployment;

/// The only two environments Pages has. A third would be silently created by a
/// `PATCH` and then read by nothing.
pub const CLOUDFLARE_ENVIRONMENTS: [&str; 2] = ["production", "preview"];

pub struct CloudflareActions {
    token: String,
    account_id: String,
}

impl CloudflareActions {
    pub fn new(token: String, account_id: String) -> Self {
        Self { token, account_id }
    }

    /// Rebuild an existing deployment.
    ///
    /// Cloudflare calls this "retry" and mints a new deployment id, inheriting
    /// the original's branch, commit and environment — so, unlike Vercel's
    /// redeploy, nothing has to be carried through by the caller to keep a
    /// production retry in production.
    pub async fn retry(
        &self,
        project: &str,
        deployment_id: &str,
    ) -> Result<CreatedDeployment, CloudflareApiError> {
        let created = self
            .post_deployment(project, deployment_id, "retry")
            .await?;
        Ok(CreatedDeployment {
            id: result_string(&created, "id").unwrap_or_default(),
            url: result_field(&created, "url"),
        })
    }

    /// Point the project's production alias at an earlier deployment without
    /// rebuilding it. Cloudflare records this as a rollback, and it takes
    /// effect immediately.
    ///
    /// Falls back to the id it was *given* when Cloudflare does not name one:
    /// a rollback does not create a deployment, so the answer describes the one
    /// production now serves.
    pub async fn rollback(
        &self,
        project: &str,
        deployment_id: &str,
    ) -> Result<CreatedDeployment, CloudflareApiError> {
        let rolled = self
            .post_deployment(project, deployment_id, "rollback")
            .await?;
        Ok(CreatedDeployment {
            id: result_string(&rolled, "id").unwrap_or_else(|| deployment_id.to_string()),
            url: result_field(&rolled, "url"),
        })
    }

    async fn post_deployment(
        &self,
        project: &str,
        deployment_id: &str,
        action: &str,
    ) -> Result<Value, CloudflareApiError> {
        let path = format!(
            "{}/deployments/{}/{action}",
            self.project_path(project),
            urlencoding::encode(deployment_id)
        );
        request_json(&self.token, "POST", &path, None).await
    }

    /// Add a variable. `encrypted` is the default, and its value never reads
    /// back over the API afterwards.
    pub async fn create_env(
        &self,
        project: &str,
        key: &str,
        value: &str,
        environments: &[String],
        secret: bool,
    ) -> Result<CloudflareEnvVar, CloudflareApiError> {
        let kind = if secret { SECRET_TEXT } else { PLAIN_TEXT };
        let mut patch = Map::new();
        for environment in require_environments(environments)? {
            patch.insert(
                environment,
                serde_json::json!({ key: { "type": kind, "value": value } }),
            );
        }
        self.patch_env(project, key, patch).await
    }

    /// Change a variable's value and/or the environments it applies to.
    ///
    /// Adding an environment needs a value: Cloudflare stores the variable
    /// separately per environment and never hands back a secret's value, so
    /// there is nothing to copy across. Saying that is better than writing an
    /// empty string into the new environment.
    pub async fn update_env(
        &self,
        project: &str,
        key: &str,
        value: Option<&str>,
        environments: Option<&[String]>,
    ) -> Result<CloudflareEnvVar, CloudflareApiError> {
        let current = self.current_env(project, key).await?;
        let wanted = match environments {
            Some(environments) => require_environments(environments)?,
            None => current.environments.clone(),
        };

        let mut patch = Map::new();
        for environment in &wanted {
            match value {
                Some(value) => {
                    patch.insert(
                        environment.clone(),
                        serde_json::json!({ key: { "type": current.kind, "value": value } }),
                    );
                }
                None if !current.environments.contains(environment) => {
                    return Err(local(format!(
                        "Adding \"{key}\" to {environment} needs a value — Cloudflare stores it separately per environment."
                    )));
                }
                None => {}
            }
        }
        // Every environment the variable is leaving, cleared by name.
        for environment in &current.environments {
            if !wanted.contains(environment) {
                patch.insert(environment.clone(), serde_json::json!({ key: Value::Null }));
            }
        }
        self.patch_env(project, key, patch).await
    }

    /// Remove a variable from every environment it is set in.
    pub async fn delete_env(&self, project: &str, key: &str) -> Result<(), CloudflareApiError> {
        let current = self.current_env(project, key).await?;
        let mut patch = Map::new();
        for environment in current.environments {
            patch.insert(environment, serde_json::json!({ key: Value::Null }));
        }
        self.patch_project(project, patch).await?;
        Ok(())
    }

    /// The one write to a project's variables, plus the check that Cloudflare
    /// actually reported the key afterwards — a `PATCH` it accepted but did not
    /// apply would otherwise read as a successful save.
    async fn patch_env(
        &self,
        project: &str,
        key: &str,
        patch: Map<String, Value>,
    ) -> Result<CloudflareEnvVar, CloudflareApiError> {
        self.patch_project(project, patch)
            .await?
            .into_iter()
            .find(|variable| variable.key == key)
            .ok_or_else(|| {
                local(format!(
                    "Cloudflare did not report \"{key}\" after the update."
                ))
            })
    }

    async fn patch_project(
        &self,
        project: &str,
        patch: Map<String, Value>,
    ) -> Result<Vec<CloudflareEnvVar>, CloudflareApiError> {
        let mut configs = Map::new();
        for (environment, variables) in patch {
            configs.insert(environment, serde_json::json!({ "env_vars": variables }));
        }
        let body = serde_json::json!({ "deployment_configs": configs });
        let updated = request_json(
            &self.token,
            "PATCH",
            &self.project_path(project),
            Some(&body),
        )
        .await?;
        Ok(merge_env_vars(
            updated
                .get("result")
                .and_then(|result| result.get("deployment_configs")),
        ))
    }

    /// What the project holds for this key right now, which both `update` and
    /// `delete` need in order to know which environments to touch.
    async fn current_env(
        &self,
        project: &str,
        key: &str,
    ) -> Result<CloudflareEnvVar, CloudflareApiError> {
        let project = request_json(&self.token, "GET", &self.project_path(project), None).await?;
        merge_env_vars(
            project
                .get("result")
                .and_then(|result| result.get("deployment_configs")),
        )
        .into_iter()
        .find(|variable| variable.key == key)
        .ok_or_else(|| local(format!("No variable named \"{key}\" on this project.")))
    }

    fn project_path(&self, project: &str) -> String {
        format!(
            "/accounts/{}/pages/projects/{}",
            self.account_id,
            urlencoding::encode(project)
        )
    }
}

/// Rejects anything that is not one of Pages' two environments.
///
/// Vercel's env dialog offers `development` as well, and the two providers
/// share that dialog — so a `PATCH` naming it would quietly create a third
/// deployment config that nothing ever reads.
fn require_environments(environments: &[String]) -> Result<Vec<String>, CloudflareApiError> {
    let invalid: Vec<&str> = environments
        .iter()
        .filter(|environment| !CLOUDFLARE_ENVIRONMENTS.contains(&environment.as_str()))
        .map(String::as_str)
        .collect();
    if !invalid.is_empty() {
        return Err(local(format!(
            "Cloudflare Pages only has {} and {} environments, not {}.",
            CLOUDFLARE_ENVIRONMENTS[0],
            CLOUDFLARE_ENVIRONMENTS[1],
            invalid.join(", ")
        )));
    }
    Ok(environments.to_vec())
}

/// A string inside Cloudflare's `result` envelope, if it sent one.
fn result_string(payload: &Value, key: &str) -> Option<String> {
    payload
        .get("result")
        .and_then(|result| result.get(key))
        .and_then(Value::as_str)
        .map(str::to_string)
}

/// A field inside `result`, narrowed to null when absent — the shape the
/// dashboard reads, where a deployment with no hostname yet is a real state.
fn result_field(payload: &Value, key: &str) -> Value {
    payload
        .get("result")
        .and_then(|result| result.get(key))
        .filter(|value| !value.is_null())
        .cloned()
        .unwrap_or(Value::Null)
}

/// A refusal this module reached on its own, with no request behind it.
fn local(message: String) -> CloudflareApiError {
    CloudflareApiError { message, status: 0 }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_pages_two_environments_are_accepted() {
        let ok = require_environments(&["production".into(), "preview".into()]).unwrap();
        assert_eq!(ok, vec!["production".to_string(), "preview".to_string()]);

        let refused = require_environments(&["production".into(), "development".into()])
            .unwrap_err()
            .message;
        // The message names what was wrong, not just that something was.
        assert!(refused.contains("development"), "{refused}");
        assert!(refused.contains("production and preview"), "{refused}");
    }

    #[test]
    fn an_empty_environment_list_is_not_this_functions_refusal() {
        // The route requires at least one before it gets here; this only
        // decides whether the *names* are ones Pages has.
        assert_eq!(require_environments(&[]).unwrap(), Vec::<String>::new());
    }
}
